### PR TITLE
[feat] Add LoRA

### PR DESCRIPTION
**Pull request by rank-Yu**
This PR introduces LoRA support.

Here's the test script used:

```bash
#!/bin/bash

set -x

MODEL_PATH=Qwen/Qwen3-0.6B  # replace it with your local file path

python3 -m verl.trainer.main \
    config=examples/config.yaml \
    worker.actor.model.model_path=${MODEL_PATH} \
    trainer.experiment_name=qwen3_0.6b_math_grpo_lora \
    worker.rollout.n=4 \
    worker.rollout.tensor_parallel_size=1 \
    trainer.total_epochs=4 \
    trainer.n_gpus_per_node=1 \
    worker.actor.model.lora.rank=16 \
    worker.actor.model.lora.alpha=16
```

This run is over a 4090, which observes acc raise:

da2822ea-5d01-4cb2-acec-b4f8bb4e2a61

### CODE REVIEW

**Comment by hiyouga** [comment]
@Kuangdd01 Could you please verify the performance on a 7B model?

**Comment by Kuangdd01** [comment]
🫡 Okay. will report result with rank=128 lr=1e-5 on Qwen2.5-7B-Inst here.

**Comment by Kuangdd01** [comment]
THX AND HAPPY NEW YEAR!
I encountered an issue when do lora tuning with multiple gpus.
E,g. when do load weights in FSDPVLLMShardingManager, some weight tensors seemed not be `DTensor` class with full_tensor() method.
``` python
def _make_weight_iterator(
    self, actor_weights: dict[str, Union[torch.Tensor, DTensor]]
) -> Iterable[tuple[str, torch.Tensor]]:
    for name, tensor in actor_weights.items():
        yield name, tensor.full_tensor() if self.world_size != 1 else tensor
```
Although i don't know what happend here, I add a if-condition to avoid it. Can @rank-Yu explain what happened here?
```python
def _make_weight_iterator(
    self, actor_weights: dict[str, Union[torch.Tensor, DTensor]]
) -> Iterable[tuple[str, torch.Tensor]]:
    items = actor_weights.items() if isinstance(actor_weights, dict) else actor_weights
    for name, tensor in items:
        if not isinstance(name, str):
            name = str(name)

        if hasattr(tensor, "full_tensor"):
            weight = tensor.full_tensor()
        else:
            weight = tensor  # torch.Tensor

        yield name, weight
```

**Comment by hiyouga** [comment]
@Kuangdd01 @rank-Yu we need to first apply lora model then do fsdp sharding to have all params in dtensor

**Comment by rank-Yu** [comment]
> THX AND HAPPY NEW YEAR! I encountered an issue when do lora tuning with multiple gpus. E,g. when do load weights in FSDPVLLMShardingManager, some weight tensors seemed not be `DTensor` class with full_tensor() method.
> 
> ```python
> def _make_weight_iterator(
>     self, actor_weights: dict[str, Union[torch.Tensor, DTensor]]
> ) -> Iterable[tuple[str, torch.Tensor]]:
>     for name, tensor in actor_weights.items():
>         yield name, tensor.full_tensor() if self.world_size != 1 else tensor
> ```
> 
> Although i don't know what happend here, I add a if-condition to avoid it. Can @rank-Yu explain what happened here?
> 
> ```python
> def _make_weight_iterator(
>     self, actor_weights: dict[str, Union[torch.Tensor, DTensor]]
> ) -> Iterable[tuple[str, torch.Tensor]]:
>     items = actor_weights.items() if isinstance(actor_weights, dict) else actor_weights
>     for name, tensor in items:
>         if not isinstance(name, str):
>             name = str(name)
> 
>         if hasattr(tensor, "full_tensor"):
>             weight = tensor.full_tensor()
>         else:
>             weight = tensor  # torch.Tensor
> 
>         yield name, weight
> ```


Thanks! This can happen because `actor_weights` is not guaranteed to be all `DTensor` even when `world_size != 1`.

In our LoRA path (e.g. `_collect_lora_params()`), we often `detach().cpu()` the tensors or materialize them during `summon_full_params`, so the resulting weights can be plain `torch.Tensor` (typically on CPU) rather than `DTensor`. In that case, gating on `world_size` and calling `.full_tensor()` will fail.

So instead of using `world_size` as the condition, I updated the iterator to only materialize when the tensor is actually a `DTensor`, and also move it to the current CUDA device before `full_tensor()` to better support FSDP + CPU offload:

```python
def _make_weight_iterator(
    self, actor_weights: dict[str, Union[torch.Tensor, DTensor]]
) -> Iterable[tuple[str, torch.Tensor]]:
    device = torch.device("cuda", torch.cuda.current_device())
    for name, tensor in actor_weights.items():
        yield name, tensor.to(device, non_blocking=True).full_tensor() if isinstance(tensor, DTensor) else tensor
```

Your `hasattr(full_tensor)` workaround is also safe; I just kept the check stricter (`isinstance(DTensor)`) and aligned device placement for the DTensor case.

File: verl/workers/actor/config.py [hunk 1]
```diff
@@ -39,6 +53,8 @@ def post_init(self):
         if self.tokenizer_path is not None and os.path.exists(self.tokenizer_path):
             self.tokenizer_path = os.path.abspath(self.tokenizer_path)
 
+        self.lora.post_init()
```
Comment (hiyouga) [inline verl/workers/actor/config.py]: we have https://github.com/hiyouga/EasyR1/blob/main/verl/trainer/config.py#L26

File: verl/workers/fsdp_workers.py [hunk 2]
```diff
@@ -624,6 +657,15 @@ def compute_log_probs(self, data: DataProto):
     def compute_ref_log_probs(self, data: DataProto):
         assert self._has_ref
 
+        if self._is_lora:
+            # if _is_lora, actor without lora applied is the ref
+            data.meta_info["is_lora"] = True
+            data = self.compute_log_probs(data)
```
Comment (hiyouga) [inline verl/workers/fsdp_workers.py]: Remember to disable the initialization of ref model to reduce VRAM consumption. And how about set the self.ref_fsdp_module as the main model to avoid invoke `self.compute_log_probs` here?

**Comment by Kuangdd01** [comment]
Qwen2.5-7B-Instruct Lora experiment [log](https://wandb.ai/kingsley/EasyR1_public/reports/EasyR1-LoRA-Run--VmlldzoxNTU5MTAwOA?accessToken=2zzjdw0doy24xltvbq0c76r7p8nz0ucjw371ynka2qi82z092ywy20zy79cycdax).

File: verl/workers/sharding_manager/fsdp_vllm.py [hunk 3]
```diff
@@ -89,17 +98,98 @@ def _make_weight_iterator(
...
+                model = self.module._fsdp_wrapped_module.base_model.model
+                orig_dev = next(model.parameters()).device
+                model = model.to("cpu")
+                for name, param in model.state_dict().items():
+                    if any(x in name for x in ["_flat_param", "lora_"]):
+                        continue
+                    name = name.replace("_fsdp_wrapped_module.", "").replace(".base_layer", "")
+                    lora_params[name] = param.detach().cpu()
+                model = model.to(orig_dev)
+        return lora_params
```
[the remark this answers is not in the document]
Comment (rank-Yu) [reply verl/workers/sharding_manager/fsdp_vllm.py]: `fsdp_version` has been removed.

File: verl/utils/vllm_utils.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                Patch vllm.model_executor.models.qwen3_vl.Qwen3VLForConditionalGeneration.get_mm_mapping in vLLM 0.11.0
+                Reason:
+                vLLM 0.11.0 uses "model.visual.*" prefixes for Qwen3-VL, but the real module names are "visual.*".
+                This breaks LoRA filtering for multimodal parts, so we align the prefixes to the real module names.
+                """
+                return MultiModelKeys.from_string_field(
+                    language_model="language_model",
+                    connector="visual.merger.",
+                    tower_model="visual.",
+                )
```
Comment (hiyouga) [inline verl/utils/vllm_utils.py]: the module name of qwen3vl is actually model.visual.*:
https://huggingface.co/Qwen/Qwen3-VL-4B-Instruct/blob/main/model.safetensors.index.json#L404-L411

**Review by hiyouga — APPROVED** [review]
LGTM!

**Review by hiyouga — APPROVED** [review]

**Comment by xlg-go** [comment]
Amazing!!!!!!!  Excellent work~~~
@rank-Yu 
@hiyouga 

Starting from vllm 0.14.0, support for LoRa VIT has been added.
https://github.com/hiyouga/EasyR1/blob/938d4eae9c38f95352cf80e8d5f107fa02dd402a/examples/config.yaml#L52

**Comment by hiyouga** [comment]
@xlg-go Thanks for the information! We'll upgrade to vllm 0.14.0 recently

**Comment by xlg-go** [comment]
> @xlg-go Thanks for the information! We'll upgrade to vllm 0.14.0 recently

Nice work!!!!!!!
Leading verl~~~~~
